src/ioctl_tree.c

```diff
@@ -357,6 +357,9 @@ ioctl_tree_execute(ioctl_tree * tree, ioctl_tree * last, IOCTL_REQUEST_TYPE id,
     ioctl_tree *i;
     int r, handled;
 
+    /* initialize return code */
+    *ret = 0;
```

Comment:
This "succeed by default" semantics is bad, as it hides errors. src/umockdev-ioctl.vala initializes the `ret` value to -1, so this is also inconsistent with the current behaviour.

Let's initialize to -1 here,  and drop the `else` in line 376 instead.

Let's also add an assertion that `ret` is not NULL, to document that requirement and make errors clearer.

Reply:
... and also drop the then-redundant initialization in umockdev-ioctl.vala